processor/spanmetricsprocessor/processor.go

```diff
@@ -231,21 +244,40 @@ func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) e
 // writes the raw metrics data into the metrics object.
 func (p *processorImp) buildMetrics() *pdata.Metrics {
 	m := pdata.NewMetrics()
-	ilm := m.ResourceMetrics().AppendEmpty().InstrumentationLibraryMetrics().AppendEmpty()
-	ilm.InstrumentationLibrary().SetName("spanmetricsprocessor")
+	rms := m.ResourceMetrics()
+	for key := range p.resourceAttrList {
+		p.lock.RLock()
+		resourceAttrKey := resourceKey(key)
+		resourceAttributesMap := p.resourceKeyToDimensions[resourceAttrKey]
+
+		// if the service name doesn't exist, we treat it as invalid and do not generate a trace
+		if _, ok := resourceAttributesMap.Get(serviceNameKey); !ok {
+			continue
+		}
+
+		rm := rms.AppendEmpty()
+
+		resourceAttributesMap.Range(func(k string, v pdata.AttributeValue) bool {
+			rm.Resource().Attributes().Insert(k, v)
+			return true
+		})
 
-	p.lock.RLock()
-	p.collectCallMetrics(ilm)
-	p.collectLatencyMetrics(ilm)
-	p.lock.RUnlock()
+		ilm := rm.InstrumentationLibraryMetrics().AppendEmpty()
+		ilm.InstrumentationLibrary().SetName("spanmetricsprocessor")
+
+		// build metrics per resource
+		p.collectCallMetrics(ilm, resourceAttrKey)
+		p.collectLatencyMetrics(ilm, resourceAttrKey)
+		p.lock.RUnlock()
+	}
 
 	return &m
 }
 
 // collectLatencyMetrics collects the raw latency metrics, writing the data
 // into the given instrumentation library metrics.
-func (p *processorImp) collectLatencyMetrics(ilm pdata.InstrumentationLibraryMetrics) {
-	for key := range p.latencyCount {
+func (p *processorImp) collectLatencyMetrics(ilm pdata.InstrumentationLibraryMetrics, resAttrKey resourceKey) {
+	for mKey := range p.latencyCount[resAttrKey] {
```

Comment:
I am not certain but you may have to hold the read lock for this loop since it could change underneath your feet.

Reply:
The lock is held in the code that calls this function so it should be fine right?